### ISSUE TITLE
decompilation of ARMv7 post-indexed strb.w instruction

### ISSUE THREAD

**Issue by Delors** [NationalSecurityAgency/ghidra#1995]
I'm currently disassembling some firmware, when I stumbled across the following code snippet produced by Ghidra (the names are already my own ones):

```asm
void memset(byte *addr,byte value,int count)
                            assume LRset = 0x0
                            assume TMode = 0x1
  undefined         r0:1           <RETURN>                                
  byte *            r0:4           addr                                    
  byte              r1:1           value
  int               r2:4           count
  undefined4        r0:4           iPtr                                    

22 b1           cbz        count,LAB_FIN
02 44           add        count,addr
                      LAB_LOOP
00 f8 01 1b     strb.w     value,[iPtr],#0x1
90 42           cmp        iPtr,count
fb d1           bne        LAB_LOOP
                      LAB_FIN  
70 47           bx         lr
00              ??         00h
bf              ??         BFh
```

Ghidra's decompiler produces the following output (after setting some types):

```c
void memset(byte *addr,byte value,int count)
{
  byte *iPtr;

  if (count != 0) {
    iPtr = addr;
    do {
      iPtr = iPtr + 1;
      *iPtr = value;                // write to iPtr AFTER the pointer was increased
      iPtr = iPtr;
    } while (iPtr != addr + count);
  }
  return;
}
```

The decompiled function suggests that this function will not set the memory at the given address (addr) to the specified value, but will always start with addr+1. This, however, doesn't feel right and as far as I understand the post-indexed strb.w instruction the order of the pointer-increment and the assignment instruction in the decompiled code is wrong.

**Comment by mumbel** [comment]
Confirmed with master.  pcode for the `strb.w` looks fine

```
$U11160:4 = COPY r0
r0 = INT_ADD r0, 1:4
$U17480:1 = SUBPIECE r1, 0:4
STORE ram($U11160), $U17480
```

**Comment by emteere** [comment]
Agree, that does not look correct reviewing the Pcode of the function.  Will investigate.

**Comment by allhavebrainimplantsandmore** [comment]
/whisper NSA got any source code that's ever been written, this is a false flag, they never decompile anything.

**Comment by caheckman** [comment]
This is fixed in the master branch now.  The decompiler was giving two distinct variables the same name, which is why you see the "iPtr = iPtr;" line.  The fix will be in 9.2.  Thank you for reporting this.
